nilearn/interfaces/bids.py

```diff
@@ -158,3 +168,324 @@ def parse_bids_filename(img_path):
         else:
             reference[field] = None
     return reference
+
+
+def save_glm_to_bids(
+    model,
+    contrasts,
+    contrast_types=None,
+    out_dir='.',
+    prefix=None,
+):
+    """Save GLM results to BIDS-like files.
+
+    Parameters
+    ----------
+    model : :obj:`nilearn.glm.first_level.FirstLevelModel` or
+            :obj:`nilearn.glm.second_level.SecondLevelModel`
+        First- or second-level model from which to save outputs.
+    contrasts : :obj:`dict` of :obj:`str`-:obj:`numpy.ndarray` pairs
+        A dictionary containing contrasts. Keys are contrast names,
+        while values are arrays containing contrast weights.
+        The arrays may be 1D or 2D.
+    contrast_types : None or :obj:`dict` of :obj:`str`s, optional
+        An optional dictionary mapping some or all of the contrast names to
+        specific contrast types ('t' or 'F'). If None, all contrast types will
+        be automatically inferred based on the contrast arrays
+        (1D arrays are t-contrasts, 2D arrays are F-contrasts).
+        Keys in this dictionary must match the keys in the ``contrasts``
+        dictionary, but only those contrasts for which contrast type must be
+        explicitly set need to be included.
+        Default is None.
+    out_dir : :obj:`str`, optional
+        Output directory for files. Default is current working directory.
+    prefix : :obj:`str` or None, optional
+        String to prepend to generated filenames.
+        If a string is provided, '_' will be added to the end.
+        Default is None.
+
+    Warning
+    -------
+    The files generated by this function are a best approximation of
+    appropriate names for GLM-based BIDS derivatives.
+    However, BIDS does not currently have GLM-based derivatives supported in
+    the specification, and there is no guarantee that the files created by
+    this function will be BIDS-compatible if and when the specification
+    supports model derivatives.
+
+    Notes
+    -----
+    This function writes files for the following:
+    - Modeling software information (dataset_description.json)
+    - Model-level metadata (statmap.json)
+    - Model design matrix (design.tsv)
+    - Model design matrix figure (design.svg)
+    - Model error (stat-errorts_statmap.nii.gz)
+    - Model r-squared (stat-rSquare_statmap.nii.gz)
+    - Contrast parameter estimates (contrast-[name]_stat-effect_statmap.nii.gz)
+    - Variance of the contrast parameter estimates
+      (contrast-[name]_stat-variance_statmap.nii.gz)
+    - Contrast test statistics (contrast-[name]_stat-[F|t]_statmap.nii.gz)
+    - Contrast p- and z-values (contrast-[name]_stat-[p|z]_statmap.nii.gz)
+    - Contrast weights figure (contrast-[name]_design.svg)
+    """
+    DATA_ATTRIBUTES = [
```

Comment:
Also, not a priority as this is still WIP, but I think that this function is way too long, and would gain in clarity from a refactoring in smaller private functions handling the various file outputs.
Also, it looks like having a simple file for all BIDS interfaces might not scale well if we plan to develop our BIDS support.
It could be worth having private utility functions useful across the bids interface in a separate file like `nilearn/interfaces/bids/_utils.py`, such that this would go in something like `nilearn/interfaces/bids/glm.py`.
WDYT?

Reply:
That sounds like a good plan. I'll break it up.